### ISSUE TITLE
che sync failing with fatal error

### ISSUE THREAD

**Issue by mike-potter** [eclipse-che/che#3840]
When running che sync via:

 docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock \
                    -v ~/Projects/che:/data \
                    -v ~/Projects/sync:/sync \
                    eclipse/che sync myworkspace

it goes for a while (couple minutes) then terminates with:

INFO: INFO: (che mount): Mounting user@172.17.0.1:/projects with SSHFS
INFO: INFO: (che mount): Successfully mounted user@172.17.0.1:/projects (32797)
INFO: INFO: (che mount): Initial sync...Please wait.
---------------------------------------
!!!
!!! ERROR: Fatal error occurred (1)
!!!
---------------------------------------

The folder being synced does have a large number of files in it (it's a Drupal installation).  Seemed to work only when I had a small number of files.

The sync *does* appear to finish.  But it doesn't keep running to sync the files.  Any way to determine more specific error information?  The basic message isn't very useful.

**Reproduction Steps:**

**Che version:**    5.1.0

**OS and version:**    Mac OSX 10.10.5

**Docker version:** 1.12.5   

**Che cli.log output:**  

Fri Jan 20 21:03:02 UTC 2017
INFO: (che cli): Checking network... (hint: '--fast' skips version, network, and nightly checks)
/usr/bin/curl  "-I -k dockerhub.com -s -o /dev/null --write-out %{http_code}"
INFO: (che cli): Loading cli...
DEBUG: docker_run
DEBUG: is_boot2docker
DEBUG: has_docker_for_windows_client
DEBUG: is_initialized
DEBUG: is_initialized
DEBUG: is_initialized
DEBUG: cli_parse
DEBUG: cmd_sync
DEBUG: update_image_if_not_found
INFO: (che download): Checking for image 'eclipse/che-mount:nightly'...
found
INFO: (che mount): Starting sync process to /Users/mpotter/Projects/sync
DEBUG: docker_run

**Comment by TylerJewell** [comment]
@mike-potter - seems odd that it is crashing out on you during the sync iteration.  We launch a special container to do the work `eclipse/che-mount`.  @benoitf will have some ideas on how you can connect to this container and stream its logs to give us an indication if anything is failing inside the ongoing sync.

Do you need to sync all files?  It's possible to mount a unison profile that further restricts which folders actively need to be mounted.

**Comment by mike-potter** [comment]
Yes, I was surprised to see something that would cause the container to crash/exit. Hopefully more detailed log info will help. 

I don't need to mount all of the files, so documentation on how to exclude some folders (like /vendor) would be helpful.  However I think we need to learn the cause of the problem to make this more stable for everybody.

**Comment by TylerJewell** [comment]
@mike-potter - I am upleveling this to a bug.  We usually do not do that until we verify and reproduce it.  But this seems likely.

You can apply a unison profile to optimize with includes or excludes.  Docs on how to do that are here:
https://www.eclipse.org/che/docs/ide/sync/index.html#optimize.  Though, not to send you on a rabbit whole, I just checked these docs and they are valid for 4.x line, but haven't been updated for the 5.0 line.

While the unison profile configuration file is still identical to what you need to enter, *how* you apply it is different.  You now need to mount the folder with that profile to the CLI with `-v <path-to-profile>:/unison`.  This will get the file into the CLI and applied to your sync path.  

I've asked engineers to take a look at docs and ensure they are current.

**Comment by benoitf** [comment]
@mike-potter are you able to build locally some docker images ?

based on your trace the exception is thrown there
https://github.com/eclipse/che/blob/master/dockerfiles/mount/entrypoint.sh#L126

but I see that all traces are turned off.
so we should turn traces on to get the error. (either if you rebuild the eclipse/che-mount:nightly image or if we add some logic there to turn traces on)

BTW, maybe you have some steps to reproduce the issue. Is that the workspace you try to sync is an opensource project that I could try to clone as well on my side ?

**Comment by mike-potter** [comment]
@benoitf I can probably build some local images with good instructions. I use docker a fair amount but haven't gotten into the creation of images much. So just let me know what to do to use the latest nightly image.  

For reproducing it might be easy for you. This is just a Drupal 8 site where I am syncing the entire docroot of the site. I do this because I work both on Drupal modules as well as Drupal core so it's good to have synced access to all the files, especially for debugging.

So if you download the latest Drupal build from here: https://www.drupal.org/project/drupal and then try to sync that top-level folder you might be able to reproduce.

**Comment by benoitf** [comment]
@mike-potter I've added log output in nightlies images (eclipse/che:nightly) of the sync utility

using `--unison-verbose`at the end of the command will print you some traces 

you may first performed a pull of the cli
```
$ docker pull eclipse/che:nightly
```

here is an example on how to tun
```
docker run .... eclipse/che:nightly sync  <your-workspace-name> --unison-verbose
```
you should then see the trace
```
INFO: INFO: (che mount): Initial sync...Please wait.
INFO: using verbose mode
```

I've tested locally with a clone of git drupal and I was able to perform the sync operation without getting an error.


```
kdir: can't create directory '/root/.ssh': File exists
INFO: INFO: (che mount): Mounting user@192.168.65.2:/projects with SSHFS
INFO: INFO: (che mount): Successfully mounted user@192.168.65.2:/projects (32832)
INFO: INFO: (che mount): Initial sync...Please wait.
INFO: using verbose mode
DEBUG: Using command unison /mntssh /mnthost -batch -auto -prefer=newer
Contacting server...
Looking for changes
Warning: No archive files were found for these roots, whose canonical names are:
	/mntssh
	/mnthost
This can happen either
because this is the first time you have synchronized these roots, 
or because you have upgraded Unison to a new version with a different
archive format.  

Update detection may take a while on this run if the replicas are 
large.

Unison will assume that the 'last synchronized state' of both replicas
was completely empty.  This means that any files that are different
will be reported as conflicts, and any files that exist only on one
replica will be judged as new and propagated to the other replica.
If the two replicas are identical, then no changes will be reported.

If you see this message repeatedly, it may be because one of your machines
is getting its address from DHCP, which is causing its host name to change
between synchronizations.  See the documentation for the UNISONLOCALHOSTNAME
environment variable for advice on how to correct this.

Donations to the Unison project are gratefully accepted: 
http://www.cis.upenn.edu/~bcpierce/unison

/ drupal/themes/README.txtReconciling changes                           
dir      ---->            drupal  
Propagating updates
UNISON 2.48.4 started propagating changes at 17:24:01.94 on 25 Jan 2017
[BGN] Copying drupal from /mntssh to /mnthost
[END] Copying drupal
UNISON 2.48.4 finished propagating changes at 17:29:44.28 on 25 Jan 2017
Saving synchronizer state
Synchronization complete at 17:29:44  (1 item transferred, 0 skipped, 0 failed)
INFO: INFO: (che mount): Background sync continues every 2 seconds.
INFO: INFO: (che mount): This terminal will block while the synchronization continues.
INFO: INFO: (che mount): To stop, issue a SIGTERM or SIGINT, usually CTRL-C.
```

**Comment by mike-potter** [comment]
Thanks for the fast work on this.  I should have a chance to test it tomorrow or early next week.

**Comment by mike-potter** [comment]
I was able to pull down the nightly and create a new workspace there.  But I never got the "using verbose mode" output.  Here is the command I used:

docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock \
-v ~/Projects/che-night:/data \
-v ~/Projects/sync:/sync \
eclipse/che:nightly sync potter --unison-verbose

and, I still got the fatal error at the end.

**Comment by mike-potter** [comment]
Here is the full output 
```
% docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock \
> -v ~/Projects/che-night:/data \
> -v ~/Projects/sync:/sync \
> eclipse/che:nightly sync potter --unison-verbose
WARN: Boot2docker detected - ensure :/data is mounted to %userprofile%
INFO: (che cli): nightly - using docker 1.12.2 / boot2docker
INFO: (che download): Pulling image eclipse/che:nightly

nightly: Pulling from eclipse/che
Digest: sha256:fb7e7cd0367f8e43903628145694bd365045bd599f93281f3cccac999f577222
Status: Image is up to date for eclipse/che:nightly

INFO: (che mount): Starting sync process to /Users/mpotter/Projects/sync
cp: can't stat '/profile/default.prf': No such file or directory
INFO: INFO: (che mount): Mounting user@192.168.99.100:/projects with SSHFS
INFO: INFO: (che mount): Successfully mounted user@192.168.99.100:/projects (32787)
INFO: INFO: (che mount): Initial sync...Please wait.
---------------------------------------
!!!
!!! ERROR: Fatal error occurred (1)
!!!
---------------------------------------
```
with a long delay between the "Initial sync...Please wait" and the actual fatal error output.  As before the sync actually completed but then exited and didn't keep watching/synching.

**Comment by benoitf** [comment]
@mike-potter could you delete image eclipse/che-action:nightly it seems it is using an old one

**Comment by benoitf** [comment]
@mike-potter sorry eclipse/che-mount:nightly
The cp can't stat is not in the latest version

**Comment by mike-potter** [comment]
Seems like there might be some missing dependencies that prevented some of the images from getting updated to "nightly".  So I ran "docker images -a" and did a "docker rmi ..." on any image that referenced eclipse and/or che.

Now when I run:
```
docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock -v ~/Projects/che-night:/data eclipse/che:nightly start
```
it downloads a bunch of images but gives me the following error at the end:
```
INFO: (che start):  Starting containers...
docker_compose --file="/data/instance/docker-compose-container.yml" -p="che" up -d >> "/data/cli.log" 2>&1
Recreating 4d954aa97aa3_che

ERROR: for che  No such image: sha256:c591fda369a047e720198f7ff7726d9cd7015d6f2cf21032e76c146ce54c64db
```
I don't know how to get the image that is being requested here.  Sorry to get stuck on such simple stuff.  Don't understand why docker isn't downloading the image that it cannot find.

**Comment by TylerJewell** [comment]
That is super weird.  That indicates you have some sort of image mutability issue.  Are you sure that you deleted all of the right images with your RMI?  You can use our CLI's `rmi` command to wipe all of the images.  Or, you can run `download --force` to forcibly remove our dependency images and then repull them.

**Comment by mike-potter** [comment]
Sorry, can you be more specific on the commands to run (rmi and download).  Also, I just need to be sure not to wipe *all* of my images because I also use docker for other work purposes.

Is there a docker command to find out what image is being referenced by that "no such image" error?  If so I could try and pull it directly.

**Comment by TylerJewell** [comment]
These are Che CLI commands:
`eclipse/che:nightly download --force` will do an RMI & a pull for JUST the images Che requires.

Or you would do it in two steps as:
`eclipse/che:nightly rmi`
`eclipse/che:nightly download`

**Comment by mike-potter** [comment]
Well, this was perplexing.  Did the "download --force" option, and that worked.  But then the "start" command still gave the same error as above regarding the image not found.

I even deleted the entire ~/Projects/che-night directory that I was using in the volume mount for the start command and still no luck.

I looked at the docker-compose-container.yml file and the only referenced image is eclipse/che-server:nightly, which exists when I do "docker ps".  I noticed the regular docker-compose.yml file and ran that locally with docker-compose and it gave the exact same error about not finding the image.

So finally, I stopped all of my containers, and restarted my docker-machine.  Then it worked!  So somehow there was an image id for eclipse/che:nightly that was cached somewhere and pointed to some old image we had deleted.

I will test the che-mount stuff again now and report back.  It's definitely using the nightly and unison verbose stuff.  I just need to wait to see what happens in a while when it finishes.

**Comment by mike-potter** [comment]
Well, I'm afraid this error still isn't helpful.  Here is the result of the sync run (which still crashed at the end):
```
[END] Copying test
UNISON 2.48.4 finished propagating changes at 16:29:03.20 on 01 Feb 2017
Saving synchronizer state
Synchronization complete at 16:29:03  (5 items transferred, 0 skipped, 0 failed)
INFO: INFO: (che mount): Background sync continues every 2 seconds.
INFO: INFO: (che mount): This terminal will block while the synchronization continues.
INFO: INFO: (che mount): To stop, issue a SIGTERM or SIGINT, usually CTRL-C.
ERROR:
ERROR: Unexpected exit: Trace output saved to /Users/mpotter/Projects/che-night/cli.log.
```
Here is a tail of the cli.log file:
```
INFO: (che mount):  Starting sync process to /Users/mpotter/Projects/sync
ERROR:
ERROR: Unexpected exit: Trace output saved to /Users/mpotter/Projects/che-night/cli.log.
```

**Comment by benoitf** [comment]
@mike-potter but now it seems it is not failing at the same time

before it was right after
INFO: INFO: (che mount): Initial sync...Please wait.

now it seems the initial sync is successfully done, no ?

**Comment by mike-potter** [comment]
It is still after the initial sync...just have more verbose output now.  I can paste the entire output if you want.  But I think it's failing exactly like before.  It has *always* finished the first sync and then just crashed.  So it just doesn't keep running to look for changes to sync.

**Comment by benoitf** [comment]
OK I thought it was on first sync based on initial report
I will check how to add more logs on the second sync

**Comment by benoitf** [comment]
@mike-potter I've added new traces for background sync command in the new nightly image `eclipse/che-mount:nightly`
If you could pull this new image and retry the command

**Comment by mike-potter** [comment]
Well, this isn't easy. I couldn't just pull the latest `eclipse/che-mount:nightly`. When I did that and tried to run `sync` it downloaded the latest nightly of che, che-server, etc, and then complained my currently running che server needed to be updated.  So I stopped the che and workspace docker containers and tried again, and then ran into that exact same 
```
ERROR: for che  No such image: sha256:c591fd...
```
as before.  This means that in order to test this again I'm going to need to remove the images and restart the docker-machine again, which means recreating my che workspace, etc and also interferring with the other projects I work on in docker.

So, I won't be able to do this today.  Hopefully tomorrow or earlier next week.  I'd love to see this process improved so that it worked more smoothly when upgrading images.
